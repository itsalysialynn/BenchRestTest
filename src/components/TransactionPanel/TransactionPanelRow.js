import React from 'react'
import { object } from 'prop-types'
import { Col } from 'react-flexbox-grid'
import dateFormatHelper from '../../helpers/dateFormatHelper'
import numberConversionHelper from '../../helpers/numberConversionHelper'
import { StyledPanelTableRow, StyledColRight } from '../../styles/StyledPanel'
import { TextMuted, TextRed, TextDark } from '../../styles/StyledText'
import LabelAndData from '../Shared/LabelAndData'

export default function TransactionPanelRow({ transaction }) {
  return (
    <StyledPanelTableRow>
      <Col xs={12} sm={6} md={2} lg={2}>
        <LabelAndData label="Date">
          <TextMuted>{dateFormatHelper.default(transaction.Date)}</TextMuted>
        </LabelAndData>
      </Col>
      <Col xs={12} sm={6} md={4} lg={4}>
        <LabelAndData label="Company">
          <TextDark>{transaction.Company}</TextDark>
        </LabelAndData>
      </Col>
      <Col xs={8} sm={6} md={4} lg={4}>
        <LabelAndData label="Account">
          <TextMuted>{transaction.Ledger}</TextMuted>
        </LabelAndData>
      </Col>
      <StyledColRight className="hidden-sm" xs={4} md={2} lg={2}>
        {transaction.Amount < 0 && (
          <LabelAndData label="Amount">
            <TextRed>
              {numberConversionHelper.toCurrency(transaction.Amount)}
            </TextRed>
          </LabelAndData>
        )}
        {transaction.Amount >= 0 && (
          <LabelAndData label="Amount">
            <TextDark>
              {numberConversionHelper.toCurrency(transaction.Amount)}
            </TextDark>
          </LabelAndData>
        )}
      </StyledColRight>
      <Col className="hidden-xs hidden-md hidden-lg hidden-xl" sm={6}>
        {transaction.Amount < 0 && (
          <LabelAndData label="Amount">
            <TextRed>
              {numberConversionHelper.toCurrency(transaction.Amount)}
            </TextRed>
          </LabelAndData>
        )}
        {transaction.Amount >= 0 && (
          <LabelAndData label="Amount">
            <TextDark>
              {numberConversionHelper.toCurrency(transaction.Amount)}
            </TextDark>
          </LabelAndData>
        )}
      </Col>
    </StyledPanelTableRow>
  )
}

TransactionPanelRow.propTypes = {
  transaction: object.isRequired,
}