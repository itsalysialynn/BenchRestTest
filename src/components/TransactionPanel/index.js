import React, { Component } from 'react'
import { Col } from 'react-flexbox-grid'
import transactionService from '../../services/transactionService'
import map from 'lodash/map'
import concat from 'lodash/concat'
import round from 'lodash/round'
import sum from 'lodash/sum'
import { StyledPanel, StyledColRight, StyledPanelTableHeader, StyledPanelBody, StyledPanelBodyEmpty } from '../../styles/StyledPanel'
import TransactionPanelRow from './TransactionPanelRow'
import numberConversionHelper from '../../helpers/numberConversionHelper'
import LoadingPanel from '../Shared/LoadingPanel'

export default class TransactionPanel extends Component {
  state = {
    transactions: [],
    loading: true,
  }

  componentDidMount() {
    this.fetchTransactions(1)
  }

  render() {
    if (this.state.loading) { return <LoadingPanel /> }

    return (
      <StyledPanel>
        <StyledPanelTableHeader>
          <Col xs={6} sm={3} md={2} lg={2}>Date</Col>
          <Col xs={6} sm={6} md={4} lg={4}>Company</Col>
          <Col xs={6} sm={3} md={4} lg={4}>Account</Col>
          <StyledColRight xs={6} sm={3} md={2} lg={2}>{numberConversionHelper.toCurrency(this.calculateTotal())}</StyledColRight>
        </StyledPanelTableHeader>
        {this.hasTransactions() && (
          <StyledPanelBody>
            {map(this.state.transactions, (transaction, index) => (
              <TransactionPanelRow key={index} transaction={transaction} />
            ))}
          </StyledPanelBody>
        )}
        {!this.hasTransactions() && (
          <StyledPanelBodyEmpty>
            No transactions found.
          </StyledPanelBodyEmpty>
        )}
      </StyledPanel>
    )
  }

  fetchTransactions = (page) => {
    transactionService.get(page)
      .then((response) => {
        const currentTransactions = this.state.transactions
        this.setState({
          transactions: concat(currentTransactions, response.transactions),
        }, () => this.fetchNextPage(response.totalCount, response.page))
      })
  }

  fetchNextPage = (totalTransactions, currentPage) => {
    if (this.state.transactions.length !== totalTransactions) {
      return this.fetchTransactions(currentPage + 1)
    }
    this.setState({ loading: false })
  }

  calculateTotal = () => (
    sum(map(this.state.transactions, transaction => +transaction.Amount)) || 0
  )

  hasTransactions = () => (
    this.state.transactions.length > 0
  )
}