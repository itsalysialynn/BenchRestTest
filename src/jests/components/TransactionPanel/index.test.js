import React from 'react'
import { shallow } from 'enzyme'
import TransactionPanel from '../../../components/TransactionPanel'

/* global describe, it, expect, test, jest, test  */

describe('loading', () => {
  const component = shallow(<TransactionPanel />).setState({ loading: true })

  it('should render the loading panel', () => {
    expect(component.find('LoadingPanel').length).toEqual(1)
  })
})

describe('no transactions', () => {
  const component = shallow(<TransactionPanel />).setState({ transactions: [], loading: false  })

  it('should render the empty message', () => {
    expect(component.html()).toMatch('No transactions found.')
  })

  test('should match snapshot', () => {
    expect(component).toMatchSnapshot()
  })
})

describe('rendering transaction panel', () => {
  const state = { transactions: [ { id: 1, Amount: '-5' } ], loading: false }
  const component = shallow(<TransactionPanel />).setState(state)

  it('should render the row', () => {
    expect(component.html()).toMatch('Date')
    expect(component.html()).toMatch('Company')
    expect(component.html()).toMatch('Account')
    expect(component.html()).toMatch('($5.00)')
  })

  test('should match snapshot', () => {
    expect(component).toMatchSnapshot()
  })
})

describe('calculating the total', () => {
  const state = {
    loading: false,
    transactions: [
      { id: 1, Amount: '525.00' },
      { id: 1, Amount: '6.29' },
      { id: 1, Amount: '5.495' },
      { id: 1, Amount: '-27' },
      { id: 1, Amount: '-73.02' },
    ],
  }
  const component = shallow(<TransactionPanel />).setState(state)

  it('should add up all the transactions', () => {
    expect(component.instance().calculateTotal()).toEqual(436.765)
  })
})


