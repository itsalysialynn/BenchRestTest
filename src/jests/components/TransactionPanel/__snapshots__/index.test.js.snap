// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no transactions should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TransactionPanel />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Styled(Row)>
          <Col
            lg={2}
            md={2}
            sm={3}
            xs={6}
          >
            Date
          </Col>
          <Col
            lg={4}
            md={4}
            sm={6}
            xs={6}
          >
            Company
          </Col>
          <Col
            lg={4}
            md={4}
            sm={3}
            xs={6}
          >
            Account
          </Col>
          <Styled(Col)
            lg={2}
            md={2}
            sm={3}
            xs={6}
          />
        </Styled(Row)>,
        false,
        <styled.div>
          No transactions found.
        </styled.div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Col
              lg={2}
              md={2}
              sm={3}
              xs={6}
            >
              Date
            </Col>,
            <Col
              lg={4}
              md={4}
              sm={6}
              xs={6}
            >
              Company
            </Col>,
            <Col
              lg={4}
              md={4}
              sm={3}
              xs={6}
            >
              Account
            </Col>,
            <Styled(Col)
              lg={2}
              md={2}
              sm={3}
              xs={6}
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "Date",
              "lg": 2,
              "md": 2,
              "sm": 3,
              "xs": 6,
            },
            "ref": null,
            "rendered": "Date",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "Company",
              "lg": 4,
              "md": 4,
              "sm": 6,
              "xs": 6,
            },
            "ref": null,
            "rendered": "Company",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "Account",
              "lg": 4,
              "md": 4,
              "sm": 3,
              "xs": 6,
            },
            "ref": null,
            "rendered": "Account",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": null,
              "lg": 2,
              "md": 2,
              "sm": 3,
              "xs": 6,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "No transactions found.",
        },
        "ref": null,
        "rendered": "No transactions found.",
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Styled(Row)>
            <Col
              lg={2}
              md={2}
              sm={3}
              xs={6}
            >
              Date
            </Col>
            <Col
              lg={4}
              md={4}
              sm={6}
              xs={6}
            >
              Company
            </Col>
            <Col
              lg={4}
              md={4}
              sm={3}
              xs={6}
            >
              Account
            </Col>
            <Styled(Col)
              lg={2}
              md={2}
              sm={3}
              xs={6}
            />
          </Styled(Row)>,
          false,
          <styled.div>
            No transactions found.
          </styled.div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Col
                lg={2}
                md={2}
                sm={3}
                xs={6}
              >
                Date
              </Col>,
              <Col
                lg={4}
                md={4}
                sm={6}
                xs={6}
              >
                Company
              </Col>,
              <Col
                lg={4}
                md={4}
                sm={3}
                xs={6}
              >
                Account
              </Col>,
              <Styled(Col)
                lg={2}
                md={2}
                sm={3}
                xs={6}
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "Date",
                "lg": 2,
                "md": 2,
                "sm": 3,
                "xs": 6,
              },
              "ref": null,
              "rendered": "Date",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "Company",
                "lg": 4,
                "md": 4,
                "sm": 6,
                "xs": 6,
              },
              "ref": null,
              "rendered": "Company",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "Account",
                "lg": 4,
                "md": 4,
                "sm": 3,
                "xs": 6,
              },
              "ref": null,
              "rendered": "Account",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": null,
                "lg": 2,
                "md": 2,
                "sm": 3,
                "xs": 6,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "No transactions found.",
          },
          "ref": null,
          "rendered": "No transactions found.",
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`rendering transaction panel should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TransactionPanel />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Styled(Row)>
          <Col
            lg={2}
            md={2}
            sm={3}
            xs={6}
          >
            Date
          </Col>
          <Col
            lg={4}
            md={4}
            sm={6}
            xs={6}
          >
            Company
          </Col>
          <Col
            lg={4}
            md={4}
            sm={3}
            xs={6}
          >
            Account
          </Col>
          <Styled(Col)
            lg={2}
            md={2}
            sm={3}
            xs={6}
          >
            ($5.00)
          </Styled(Col)>
        </Styled(Row)>,
        <styled.div>
          <TransactionPanelRow
            transaction={
              Object {
                "Amount": "-5",
                "id": 1,
              }
            }
          />
        </styled.div>,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Col
              lg={2}
              md={2}
              sm={3}
              xs={6}
            >
              Date
            </Col>,
            <Col
              lg={4}
              md={4}
              sm={6}
              xs={6}
            >
              Company
            </Col>,
            <Col
              lg={4}
              md={4}
              sm={3}
              xs={6}
            >
              Account
            </Col>,
            <Styled(Col)
              lg={2}
              md={2}
              sm={3}
              xs={6}
            >
              ($5.00)
            </Styled(Col)>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "Date",
              "lg": 2,
              "md": 2,
              "sm": 3,
              "xs": 6,
            },
            "ref": null,
            "rendered": "Date",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "Company",
              "lg": 4,
              "md": 4,
              "sm": 6,
              "xs": 6,
            },
            "ref": null,
            "rendered": "Company",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "Account",
              "lg": 4,
              "md": 4,
              "sm": 3,
              "xs": 6,
            },
            "ref": null,
            "rendered": "Account",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "($5.00)",
              "lg": 2,
              "md": 2,
              "sm": 3,
              "xs": 6,
            },
            "ref": null,
            "rendered": "($5.00)",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <TransactionPanelRow
              transaction={
                Object {
                  "Amount": "-5",
                  "id": 1,
                }
              }
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "function",
            "props": Object {
              "transaction": Object {
                "Amount": "-5",
                "id": 1,
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      false,
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Styled(Row)>
            <Col
              lg={2}
              md={2}
              sm={3}
              xs={6}
            >
              Date
            </Col>
            <Col
              lg={4}
              md={4}
              sm={6}
              xs={6}
            >
              Company
            </Col>
            <Col
              lg={4}
              md={4}
              sm={3}
              xs={6}
            >
              Account
            </Col>
            <Styled(Col)
              lg={2}
              md={2}
              sm={3}
              xs={6}
            >
              ($5.00)
            </Styled(Col)>
          </Styled(Row)>,
          <styled.div>
            <TransactionPanelRow
              transaction={
                Object {
                  "Amount": "-5",
                  "id": 1,
                }
              }
            />
          </styled.div>,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Col
                lg={2}
                md={2}
                sm={3}
                xs={6}
              >
                Date
              </Col>,
              <Col
                lg={4}
                md={4}
                sm={6}
                xs={6}
              >
                Company
              </Col>,
              <Col
                lg={4}
                md={4}
                sm={3}
                xs={6}
              >
                Account
              </Col>,
              <Styled(Col)
                lg={2}
                md={2}
                sm={3}
                xs={6}
              >
                ($5.00)
              </Styled(Col)>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "Date",
                "lg": 2,
                "md": 2,
                "sm": 3,
                "xs": 6,
              },
              "ref": null,
              "rendered": "Date",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "Company",
                "lg": 4,
                "md": 4,
                "sm": 6,
                "xs": 6,
              },
              "ref": null,
              "rendered": "Company",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "Account",
                "lg": 4,
                "md": 4,
                "sm": 3,
                "xs": 6,
              },
              "ref": null,
              "rendered": "Account",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "($5.00)",
                "lg": 2,
                "md": 2,
                "sm": 3,
                "xs": 6,
              },
              "ref": null,
              "rendered": "($5.00)",
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <TransactionPanelRow
                transaction={
                  Object {
                    "Amount": "-5",
                    "id": 1,
                  }
                }
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "function",
              "props": Object {
                "transaction": Object {
                  "Amount": "-5",
                  "id": 1,
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        false,
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
